# UTS PENAMBANGAN DATA

Nama : Hamam Mubarok Al Hadad

NIM : 210411100060

Kelas : C

Pada dasarnya, KNN merupakan algoritma supervised machine learning yang digunakan untuk mencari pola pada data yang ada dengan tujuan untuk membuat prediksi. Prediksi tersebut dapat berupa memprediksi harga mobil atau mengidentifikasi pelanggan yang akan merespons kampanye pemasaran email baru dengan baik.

Konsep KNN yang sederhana adalah dengan mengidentifikasi titik data yang memiliki fitur yang mirip dengan titik data yang ingin diprediksi, kemudian menuliskannya berdasarkan kemiripan (similarity), dan memilih k jumlah akhir dari daftar tersebut. Prediksi kemudian dapat dilakukan dengan mengambil rata-rata atau mean dari fitur dari titik akhir.

Dalam implementasi algoritma KNN, kita menggunakan metrik kemiripan untuk menghitung kemiripan antara data. Metrik kemiripan digunakan untuk membandingkan satu set fitur numerik yang tetap antara dua pengamatan.

1. Import library yang diperlukan

In [ ]:
import pandas as pd
import numpy as np
from sklearn.neighbors import KNeighborsRegressor 
from sklearn.preprocessing import LabelEncoder

2. Mengambil data yang akan kita analisa

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/kb22/Heart-Disease-Prediction/master/dataset.csv')

jumlah_baris = df.shape[0]
df.head(jumlah_baris)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

3. Mengecek nilai yang hilang/kosong

In [ ]:
df_copy = df.copy()
df_copy.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

4. Ubah nilai yang terdapat tanda tanya (?) menjadi nilai kosong

In [ ]:
df_copy = df_copy.replace('?',np.nan)

5. Cek lagi berapa nilai yang kosong pada data

In [ ]:
df_copy.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

6. Untuk melakukan normalisasi data, salah satu metode yang dapat digunakan adalah metode min-max. Metode ini mengubah setiap nilai pada kolom tertentu ke dalam rentang antara 0 hingga 1 dengan menggunakan rumus (x - min) / (max - min). Terdapat juga beberapa metode normalisasi lain seperti simple feature scaling atau z-score, namun perlu diingat bahwa kolom target (dalam kasus ini, kolom species) tidak perlu dinormalisasi.

In [ ]:
target_column = df_copy['target']
df_copy = df_copy[['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']].copy()
df_copy = (df_copy - df_copy.min())/(df_copy.max() - df_copy.min())
df_copy['target'] = target_column.copy()
df_copy.head(jumlah_baris)

age  sex        cp  trestbps      chol  fbs  restecg   thalach  \
0    0.708333  1.0  1.000000  0.481132  0.244292  1.0      0.0  0.603053   
1    0.166667  1.0  0.666667  0.339623  0.283105  0.0      0.5  0.885496   
2    0.250000  0.0  0.333333  0.339623  0.178082  0.0      0.0  0.770992   
3    0.562500  1.0  0.333333  0.245283  0.251142  0.0      0.5  0.816794   
4    0.583333  0.0  0.000000  0.245283  0.520548  0.0      0.5  0.702290   
..        ...  ...       ...       ...       ...  ...      ...       ...   
298  0.583333  0.0  0.000000  0.433962  0.262557  0.0      0.5  0.396947   
299  0.333333  1.0  1.000000  0.150943  0.315068  0.0      0.5  0.465649   
300  0.812500  1.0  0.000000  0.471698  0.152968  1.0      0.5  0.534351   
301  0.583333  1.0  0.000000  0.339623  0.011416  0.0      0.5  0.335878   
302  0.583333  0.0  0.333333  0.339623  0.251142  0.0      0.0  0.786260   

     exang   oldpeak  slope    ca      thal  target  
0      0.0  0.370968    0.0  0.00  0.333333       1  
1      0.0  0.564516    0.0  0.00  0.666667       1  
2      0.0  0.225806    1.0  0.00  0.666667       1  
3      0.0  0.129032    1.0  0.00  0.666667       1  
4      1.0  0.096774    1.0  0.00  0.666667       1  
..     ...       ...    ...   ...       ...     ...  
298    1.0  0.032258    0.5  0.00  1.000000       0  
299    0.0  0.193548    0.5  0.00  1.000000       0  
300    0.0  0.548387    0.5  0.50  1.000000       0  
301    1.0  0.193548    0.5  0.25  1.000000       0  
302    0.0  0.000000    0.5  0.25  0.666667       0  

[303 rows x 14 columns]

7. Langkah terakhirnya adalah membuat objek knn dari kelas KNeighborsRegressor, yang menerapkan algoritma K-Nearest Neighbors untuk melakukan regresi. Selanjutnya, kita melakukan randomisasi indeks dari dataset dengan menggunakan np.random.permutation() dan seed 1, sehingga dapat memastikan bahwa hasil yang dihasilkan konsisten.

  > selanjutnya adalah membagi dataset menjadi data training sebesar 75% dan data testing sebesar 25% menggunakan variabel train_index. Data training akan disimpan di dalam variabel training_data, sementara data testing akan disimpan di dalam variabel testing_data.

  > Setelah melakukan pembagian dataset, kita melakukan pelatihan model dengan menggunakan data training. Hal ini dilakukan dengan memanggil method fit() pada objek knn, dengan argumen input berupa fitur-fitur yang digunakan (yaitu sepal_length, sepal_width, petal_length, dan petal_width) dan label yaitu species.

  > Setelah pelatihan model selesai, kita melakukan prediksi pada data testing dengan memanggil method predict() pada objek knn. Input yang diberikan pada method predict() adalah fitur-fitur dari data testing yang ingin diprediksi.

  > Setelah melakukan prediksi, kita menghitung akurasi model dengan mengukur berapa persen data testing yang berhasil diprediksi dengan menggunakan threshold default 0.5.

  > Terakhir, mencetak hasil akurasi model.

In [ ]:
knn = KNeighborsRegressor()

# before we split, make sure you randomize the dataset
np.random.seed(1)
index_shuffled = np.random.permutation(df_copy.index)
shuffled_car_data = df_copy.loc[index_shuffled]

# We will use 75% of the data for training and the remaining 25% for testing
train_index = int(shuffled_car_data.shape[0] * 0.75)
training_data = shuffled_car_data.iloc[0:train_index]
testing_data = shuffled_car_data.iloc[train_index:]

# fit the model with the training data
knn.fit(training_data[['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']], training_data['target'])

# predict the target of the testing data
prediction = knn.predict(testing_data[['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']])

# calculate the accuracy of the model
accuracy = np.mean(np.abs(prediction - testing_data['target']) <= 0.5)

print(f"Akurasi model KNN: {accuracy:.2%}")


Akurasi model KNN: 92.11%


KESIMPULAN :

> Dari hasil yang telah diperoleh, diketahui bahwa tingkat akurasi model KNN pada dataset iris dengan menggunakan beberapa fitur ('sepal_length', 'sepal_width', 'petal_length', 'petal_width') adalah sebesar 92.11%.